### ISSUE TITLE
[test_pow_fractional.py:49] | (Runtime) Integration Tests | runtime-integration-tests / runtime_fd_python_3 [wh_n150_civ2]

### ISSUE THREAD

**Issue by ebanerjeeTT** [tenstorrent/tt-metal#48052]
## CI Failure Report

**Summary:** [CI] test_pow_fractional_composite PCC accuracy failure in runtime_fd_python_3 [wh_n150_civ2]

**Workflow:** (Runtime) Integration Tests
**Job:** runtime-integration-tests / runtime_fd_python_3 [wh_n150_civ2]

### Regression timeline

- **Last passing commit:** N/A (not provided)
- **First failing commit:** `ae414c18e46f` — 2026-06-24T06:20:46Z — [job](https://github.com/tenstorrent/tt-metal/actions/runs/28079421656/job/83140314814)
- **Most recent failing commit:** `1a118dcad608` — 2026-06-25T06:23:30Z — [job](https://github.com/tenstorrent/tt-metal/actions/runs/28151153934/job/83377688898)

### Failing job URLs (last 2 runs)

- https://github.com/tenstorrent/tt-metal/actions/runs/28151153934/job/83377688898
- https://github.com/tenstorrent/tt-metal/actions/runs/28079421656/job/83140314814

### Failing test path(s)

- `tests/tt_eager/python_api_testing/unit_testing/misc/test_pow_fractional.py::test_pow_fractional_composite`

### Error excerpt (terminal failure from logs)

**Run 1 (commit `1a118dcad608`):**
```
>       assert passing
E       assert False

tests/tt_eager/python_api_testing/unit_testing/misc/test_pow_fractional.py:49: AssertionError

FAILED tests/tt_eager/python_api_testing/unit_testing/misc/test_pow_fractional.py::test_pow_fractional_composite - assert False
```

**Run 2 (commit `ae414c18e46f`):**
```
>       assert passing
E       assert False

tests/tt_eager/python_api_testing/unit_testing/misc/test_pow_fractional.py:49: AssertionError

FAILED tests/tt_eager/python_api_testing/unit_testing/misc/test_pow_fractional.py::test_pow_fractional_composite - assert False
```

The logged PCC value is identical in both runs: `0.6391644024467981`

### Reproduction steps

From the repo root inside a suitable environment (Wormhole N150, CIv2):

```bash
pytest tests/tt_eager/python_api_testing/unit_testing/misc/test_pow_fractional.py::test_pow_fractional_composite -xvvv
```

Full CI command (split group):
```bash
pytest tests/tt_eager/python_api_testing/unit_testing/ -xvvv --splits 7 --group 3
```

### Notes

- The test computes a composite fractional power (`x^y` decomposed as `x^floor(y) * exp(log(x) * frac(y))`) and checks PCC against a PyTorch reference. The PCC value of ~0.639 is well below the passing threshold.
- The same fixed random seed (`torch.manual_seed(577215)`) is used, which explains the identical PCC value across runs.
- No timeout or hardware hang was involved — this is a pure numerical accuracy regression.

**Determinism check:**
- Run 1 terminal error: `tests/tt_eager/python_api_testing/unit_testing/misc/test_pow_fractional.py:49: AssertionError` — `assert False` (PCC = 0.6391644024467981)
- Run 2 terminal error: `tests/tt_eager/python_api_testing/unit_testing/misc/test_pow_fractional.py:49: AssertionError` — `assert False` (PCC = 0.6391644024467981)
- Both runs: identical test, identical file/line, identical PCC value → deterministic.

`Auto-triage-workflow: (Runtime) Integration Tests`
`Auto-triage-job-name: runtime-integration-tests / runtime_fd_python_3 [wh_n150_civ2]`

## About this issue (automated CI auto-triage)
This issue was opened and assigned by an automated AI system. It posts through a bot token, so notifications may look like they come from a person, but they are automated.

The owner is selected from recent commit history and job-ownership data. It reflects who is *likely able to fix or route* this, not necessarily who caused it, so it can be imperfect.

**If this was assigned to the wrong person, you can:**
- Comment `REASSIGN OWNER: <why this is outside your area / which team or test-area it belongs to>` and the system will automatically re-route it to a better-suited owner.
- Unassign yourself (this stops notifications to you).
- Close the issue if it is not a real or relevant failure.

The assigned owner is reminded about every 3 days until the issue is resolved. Full guide: https://tenstorrent.atlassian.net/wiki/spaces/MI6/pages/2545221692/Auto-Triage+Info

## Pipeline report
Analysis report: not found
Slack notification: not found

**Comment by Aswinmcw** [comment]
## Analysis: `test_pow_fractional_composite` PCC failure

*Reproduces on both WH N150 (civ2) and BH P150b (civ2) — PCC ≈ 0.639 vs threshold 0.99*

---

### Test overview

`test_pow_fractional_composite` computes:
```python
pow_frac_composite = exp(log(x, fast_and_approximate_mode=True) * y_frac) * pow(x, y_floor)
```

The input `x = torch.randn((1, 2, 32, 32))` contains ~50% negative values. For negative `x`, `log(x)` → NaN, propagating NaN through the whole expression. The golden `torch.pow(x, y)` also returns NaN for negative non-integer powers. A PCC of 0.639 (vs 0.99) indicates **systematic error across the positive-value elements**, not just a NaN-propagation mismatch.

---

### Auto-triage first-failing-commit is a red herring

The auto-triage reports `ae414c18e46f` (June 23, rotary embedding RISC-V cleanup) as the first failing commit — but it only touches rotary embedding kernels and has no interaction with eltwise/log/exp ops. "Last passing commit: N/A" means no passing run was found in the monitoring window; the regression started before the triage window. Look earlier.

---

### Root cause candidates

#### Candidate 1 (PRIMARY) — `cec422302cbd` · "RoundMode clarification (tt_metal/hw) 3/6" · June 15

PR #46851 changed `ckernel_sfpu_log.h` for **both WH and BH**:

```cpp
// BEFORE:
e_float = sfpi::convert<sfpi::vFloat>(abs_e, sfpi::RoundMode::NearestEven);
result  = sfpi::convert<sfpi::vFloat16b>(result, sfpi::RoundMode::NearestEven);

// AFTER (current):
e_float = sfpi::convert<sfpi::vFloat>(abs_e, sfpi::RoundMode::Nearest);
result  = sfpi::convert<sfpi::vFloat16b>(result, sfpi::RoundMode::Nearest);
```

The PR is described as "backwards compatible" because WH/BH hardware rounds nearest-away-from-zero. However, SFPI was bumped 7.58.0 → 7.59.0 → 7.60.0 between June 15–17. If SFPI 7.59.0/7.60.0 emits a different hardware instruction for `RoundMode::Nearest` than for `RoundMode::NearestEven`, the actual rounding behaviour of the BF16 conversion (`result = convert<vFloat16b>(result, ...)`) changes. A 1-ULP difference in log results would not cause PCC 0.639 alone, but incorrect exponent selection in the polynomial path would.

The fast-approximate path (used by this test) also skips normalisation:

```cpp
if constexpr (!FAST_APPROX) {
    a = a * sfpi::vConst1 + sfpi::vConst0;  // flush -0.0 / subnormals → +0.0
}
```

Without normalisation, edge-case inputs interact differently with the changed RoundMode conversion.

#### Candidate 2 (SECONDARY) — `987c7451ded8` · "Remove dependency on debug bit 11 for typecast" · June 21

PR #46231 removed debug bit 11 management from `_llk_math_eltwise_sfpu_uninit_()` on both WH and BH. Previously, after every SFPU op with FP32 dest accumulation enabled, uninit re-engaged the HW bug workaround:

```cpp
// REMOVED:
if (cfg_read(ALU_ACC_CTRL_Fp32_enabled_ADDR32) & ALU_ACC_CTRL_Fp32_enabled_MASK)
{
    _llk_math_dbg_feature_enable_();  // re-enable bit-11 workaround for next op
}
```

Debug bit 11 (`RISCV_DEBUG_REG_DBG_FEATURE_DISABLE`) is the FPU HW bug workaround from issue #46219. After this PR, if any preceding operation in the same test session leaves bit 11 in the wrong state, subsequent log/exp SFPU ops run without the workaround — potentially exercising the underlying FPU bug. This commit landed June 21, two days before the first failing CI run.

---

### Suggested investigation steps

1. **Bisect `cec422302cbd`** — Revert `RoundMode::Nearest` → `RoundMode::NearestEven` in `ckernel_sfpu_log.h` (WH + BH) and re-run the test. If PCC recovers, check whether SFPI 7.60.0 emits different hardware instructions for the two enum values.

2. **Bisect `987c7451ded8`** — Restore the bit-11 re-enable in `_llk_math_eltwise_sfpu_uninit_()` and re-run. If PCC recovers, determine whether log/exp require the bit-11 workaround and whether they should have been included in the PR's scope alongside typecast and reduce.

3. **Isolate the positive-value path** — Change the test input to avoid negatives (`torch.clamp(x, min=1e-3)`) and re-run. If PCC is still low, the error is entirely in the log polynomial computation (pointing to `cec422302cbd`). If PCC recovers to near-1.0, the error was in NaN handling or propagation.

4. **Verify fast vs full approx** — Re-run with `fast_and_approximate_mode=False` to confirm whether the regression is fast-path-specific.

---

### Fix recommendation

- **If `cec422302cbd` is the cause**: revert to `RoundMode::NearestEven` in `ckernel_sfpu_log.h` for both architectures, or confirm with the SFPI team that `Nearest` and `NearestEven` produce identical instructions in SFPI 7.59.0+.
- **If `987c7451ded8` is the cause**: explicitly manage bit 11 state in the log/exp SFPU init path (similar to the old typecast handling), or file a follow-up to verify that no SFPU op other than typecast/reduce required bit 11 management.
- **Test hardening**: the test applies fast-approximate log to a tensor with ~50% negative values from an unfixed random seed. Consider adding `torch.clamp(x, min=0.001)` or using `torch.abs(x)` as a pre-condition to restrict testing to the valid domain, and separately test NaN propagation explicitly.

cc @JasonDTX (PR #46851 author) @amahmud (PR #46231 author)

**Comment by mcw-anasuya** [comment]
Same issue as discussed here https://github.com/tenstorrent/tt-metal/issues/48053#issuecomment-4808601443

**Comment by mcw-anasuya** [comment]
Fix merged to main https://github.com/tenstorrent/tt-metal/pull/48227
Closing this issue.
